src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs

```diff
@@ -844,16 +844,95 @@ public override BoundNode VisitIsPatternExpression(BoundIsPatternExpression node
-            var expressionState = ResultType;
-            var state = PossiblyConditionalState.Create(this);
-            var labelStateMap = LearnFromDecisionDag(node.Syntax, node.DecisionDag, node.Expression, expressionState, ref state);
-            var trueState = labelStateMap.TryGetValue(node.IsNegated ? node.WhenFalseLabel : node.WhenTrueLabel, out var s1) ? s1.state : UnreachableState();
-            var falseState = labelStateMap.TryGetValue(node.IsNegated ? node.WhenTrueLabel : node.WhenFalseLabel, out var s2) ? s2.state : UnreachableState();
-            labelStateMap.Free();
-            SetConditionalState(trueState, falseState);
+            if (!VisitPossibleConditionalAccess(node.Expression, out var conditionalStateWhenNotNull)
+                || !trySetStateWhenNotNull(node.Pattern, conditionalStateWhenNotNull))
+            {
+                var expressionState = ResultType;
+                var state = PossiblyConditionalState.Create(this);
+                var labelStateMap = LearnFromDecisionDag(node.Syntax, node.DecisionDag, node.Expression, expressionState, ref state);
+                var trueState = labelStateMap.TryGetValue(node.IsNegated ? node.WhenFalseLabel : node.WhenTrueLabel, out var s1) ? s1.state : UnreachableState();
+                var falseState = labelStateMap.TryGetValue(node.IsNegated ? node.WhenTrueLabel : node.WhenFalseLabel, out var s2) ? s2.state : UnreachableState();
+                labelStateMap.Free();
+                SetConditionalState(trueState, falseState);
+            }
             SetNotNullResult(node);
             return null;
+
+            bool trySetStateWhenNotNull(BoundPattern pattern, PossiblyConditionalState conditionalStateWhenNotNull)
+            {
+                Debug.Assert(!IsConditionalState);
+
+                if (conditionalStateWhenNotNull.IsConditionalState)
+                {
+                    switch (getMatchingBoolValues(pattern))
+                    {
+                        case (hasTrue: true, hasFalse: false, hasNull: false):
+                            SetConditionalState(conditionalStateWhenNotNull.StateWhenTrue, State);
+                            return true;
+                        case (hasTrue: false, hasFalse: true, hasNull: false):
+                            SetConditionalState(conditionalStateWhenNotNull.StateWhenFalse, State);
+                            return true;
+                        case (hasTrue: false, hasFalse: true, hasNull: true):
+                            SetConditionalState(State, conditionalStateWhenNotNull.StateWhenTrue);
+                            return true;
+                        case (hasTrue: true, hasFalse: false, hasNull: true):
+                            SetConditionalState(State, conditionalStateWhenNotNull.StateWhenFalse);
+                            return true;
+                    }
+                }
+
+                switch (IsTopLevelNonNullTest(node.Pattern))
```

Comment:
If the `hasNull` flag from `getMatchingBoolValues` was a bool?, then it would strictly more expressive than `IsTopLevelNonNullTest`. Why check the pattern again?
Then we'd just add two extra cases: `(_, _, true)` and `(_, _, false)`.

Reply:
> Why check the pattern again?

With the recently pushed changes we no longer do this. Thanks